### PR TITLE
Improve client documentation

### LINKED ISSUE #803: Document thread-safety of GCS operations
The `google::cloud::storage::Client` member functions should document if it is safe to use them from multiple threads.  Consider the example of:

https://en.cppreference.com/w/cpp/container

### PR DESCRIPTION
**Pull request by coryan**
Several fixes to the `google::cloud::storage::Client` class documentation.

First, the top-level description of the class no longer warns that it is an incomplete class, that is no longer true (or at least it is more false than true). The description says how expensive it is to create the class, and the thread-safety guarantees of the class (fixes #803).

Second, the operations are grouped by resource, Doxygen sorts alphabetically by default which makes it hard to see how the functions relate to each other.

Finally, fixed a small TODO where `CopyObject` was supposed to reference `RewriteObject` as a better alternative.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1215)

### CODE REVIEW

File: google/cloud/storage/client.h [hunk 1]
```diff
@@ -1288,7 +1386,20 @@ class Client {
...
     return raw_client_->PatchObjectAcl(request).second;
   }
+  //@}
 
+  //@{
+  /**
+   * @name Bucket Default Object Access Control List operations.
+   *
+   * When you upload an object to GCS, without specifying an ACL the object is
+   * created with the Default Object ACL for that bucket. These operations are
```
Comment (houglum) [inline google/cloud/storage/client.h]: s/for that/for its/

File: google/cloud/storage/client.h [hunk 2]
```diff
@@ -1090,7 +1168,27 @@ class Client {
...
+   * @name Object Access Control List operations.
+   *
+   * You can control who has access to your GCS buckets and objects as
+   * well as what level of access they have.
+   *
+   * @note When you upload (create) an object the entity that uploads the object
+   *     is listed as the object owner. This can be a user or a service account,
+   *     depending on what credentials are used to authenticate with GCS.
+   *     Object ownership cannot be changed by modifying ACLs. You can change
+   *     object ownership only by overwriting the object.
```
Comment (houglum) [inline google/cloud/storage/client.h]: Nit: maybe clarify and change this to "by overwriting the object while authenticated as the desired owner."

File: google/cloud/storage/client.h [hunk 3]
```diff
@@ -908,7 +959,34 @@ class Client {
...
+   * well as what level of access they have.
+   *
+   * @note In most cases, you should use IAM permissions instead of ACLs to
+   * control access to buckets.
+   *
+   * @note All buckets are owned by the project owners group. Project owners
+   *     are granted `OWNER` permissions to all buckets inside their project.
+   *     Bucket and object ownership cannot be changed by modifying ACLs. If you
+   *     apply a new ACL to a bucket, be sure that the bucket owner remains
+   *     unchanged in the new ACL.
```
Comment (houglum) [inline google/cloud/storage/client.h]: I'm confused as to why this last sentence is present.  If users can't change the owner via ACL operations, whatever they do in the ACL update operation will have no effect on the original OWNER-role entry.  Maybe a more accurate wording might be something like what we have in the gsutil docs: "You cannot remove the project owners group from a bucket ACL. Attempts to do so will appear to succeed, but the service will add the project owners group into the new  ACL before applying it."

File: google/cloud/storage/client.h [hunk 4]
```diff
@@ -68,6 +91,18 @@ class Client {
     return raw_client_;
   }
 
+  //@{
+  /**
+   * @name Bucket operations.
+   *
+   * Buckets are the basic containers that hold your data. Everything that you
```
Comment (houglum) [inline google/cloud/storage/client.h]: s/Everything/Every object/

Comment (coryan) [reply google/cloud/storage/client.h]: But also notifications are associated with buckets, and the wording is verbatim from here:

https://cloud.google.com/storage/docs/key-terms#buckets

File: google/cloud/storage/client.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/client.h]: I agree.  Removed the old wording and used yours.  Note that I took that wording more or less verbatim from here:

https://cloud.google.com/storage/docs/access-control/lists#bestpractices

File: google/cloud/storage/client.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/client.h]: Done.

File: google/cloud/storage/client.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed.

File: google/cloud/storage/client.h [hunk 5]
```diff
@@ -1288,7 +1386,20 @@ class Client {
     request.set_multiple_options(std::forward<Options>(options)...);
     return raw_client_->PatchObjectAcl(request).second;
   }
+  //@}
 
+  //@{
+  /**
+   * @name Bucket Default Object Access Control List operations.
+   *
+   * When you upload an object to GCS, without specifying an ACL the object is
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed, thanks!

File: google/cloud/storage/client.h (hunk 4 above)
Comment (houglum) [reply google/cloud/storage/client.h]: Seems reasonable enough, and it's consistent with other wording, so LGTM :)

**Pull request merged**